### PR TITLE
Move form extension parsing to frontend

### PR DESCRIPTION
**Pull request by tamarasaurus**
**Description (for Contributor and Core Developer)**

The goal of this PR is to extract the processing of form extensions from the backend. This will allow the following improvements:
- Decouple from backend
- No need to clear app cache every time the extension yml files are changed
- Webpack can watch for changes in the yml files and automatically rebuild the frontend

The way it works in this PR, there is one script that collects and merges all the form extension config ymls - `form-extensions.js`, and then webpack injects the config into the `config-provider` module. The `config-provider` then filters the extensions by whether or not the ACL is granted (just as the FormExtensionProvider did before)

Tasks
- [ ] Implement the same in EE
- [x] Move form extension parsing to node
- [x] Generate json file containing the data that webpack needs (based on form extensions, config, requirejs.yml) instead of doing it in memory in webpack.config.js
- [x] Add a hook in webpack watch before recompilation to regenerate this file 
- [x] Refactor webpack config to allow form extension ymls to be watched
- [x] Upgrade to the latest webpack (4.4.1)
- [ ] Add js update detection to watch rebuild
- [ ] Add LiveReload trigger to watch rebuild
- [ ] Write docs for how it works, how to extend
- [ ] Remove the need for `pim:installer:dump-require-paths`

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: webpack/generate-manifest.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    aliases = Object.assign(aliases, {
+        'require-polyfill': resolve(__dirname, './require-polyfill.js'),
+        'require-context': resolve(__dirname, './require-context.js'),
+        'module-registry': resolve(rootDir, './web/js/module-registry.js'),
+        routes: resolve(rootDir, './web/js/routes.js'),
+        'fos-routing-base': resolve(rootDir, './vendor/friendsofsymfony/jsrouting-bundle/Resources/public/js/router.js'),
+        summernote: resolve(rootDir, 'node_modules/summernote/dist/summernote.js'),
+        json: resolve(rootDir, 'node_modules/JSON2/json2.js'),
+        jquery: require.resolve('jquery'),
```
Comment (tamarasaurus) [inline webpack/generate-manifest.js]: resolve natively instead

File: webpack/generate-manifest.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
+const process = require('process');
+const requirePaths = require('../web/js/require-paths');
```
Comment (tamarasaurus) [inline webpack/generate-manifest.js]: move need for this file

File: webpack/watch.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+const process = require('process');
+const nodemon = require('nodemon');
+const manifest = require('./manifest.json');
+
+nodemon({
+    watch: manifest.extensionPaths,
+    exec: 'yarn run webpack-dev',
```
Comment (tamarasaurus) [inline webpack/watch.js]: add commands dependent on filename of changed file

File: package.json [hunk 4]
```diff
@@ -4,9 +4,10 @@
   "description": "Akeneo PIM Application",
   "homepage": "http://www.akeneo.com",
   "scripts": {
-    "webpack": "webpack --config webpack.config.js --env=prod",
-    "webpack-dev": "webpack --config webpack.config.js",
-    "webpack-watch": "webpack --progress --config webpack.config.js --watch",
+    "webpack": "node ./webpack/generate-manifest.js && webpack --config webpack.config.js --env=prod --mode=production",
+    "webpack-dev": "node ./webpack/generate-manifest.js && webpack --config webpack.config.js --mode=development",
+    "webpack-watch": "node ./webpack/generate-manifest.js && webpack --progress --config webpack.config.js --watch --mode=development",
```
Comment (tamarasaurus) [inline package.json]: replace with the below

**Comment by jjanvier** [comment]
And this is also the kind of "big change" that seems more suitable for a major version imo.

**Comment by tamarasaurus** [comment]
@jjanvier Thanks! So far there hasn't been an impact to the customer except that they will have to update their dependencies in package.json (which is automatic). The commands are exactly the same, and the backend still tells webpack which bundles are registered in order to compile the js modules. They don't have to change anything about their workflow unless they have overridden/extended the FormExtensionProvider. For me I don't see it as a big change, but I don't know all the edge cases yet

**Comment by aRn0D** [comment]
I agree with JJ for the target version but this PR is really cool :+1:

**Comment by phaseinducer** [comment]
Agree on the 3.0 target too. There's no emergency.
Really nice to see this :+1: The fact that you burn a bunch of useless stuff in the backend is a good sign that it's the way to go.

**Comment by tamarasaurus** [comment]
In that case If we make 3.0 the target I will be a bit more ruthless in the decoupling ;) It would be cool to brainstorm what other improvements we can make

**Comment by juliensnz** [comment]
@jjanvier we can discuss about it but this doesn't change anything to external people. This is purely internal and doesn't change anything from the outside. Also, we may do some feature flag for the next subjects and start to work on 3.0 subject on the 2.3 branch (on a dedicated folder/bundle). In this case, we will need this on the 2.3 branch. 

This PR could benefit to the 2.3 with no costs and avoid to make bigger the already big gap between the 2.3 and 3.0.

**Pull request closed without merging**